File: newelixirideaboard/lib/idea_board/services/ideas_enrichment_service.ex
defmodule IdeaBoard.IdeasEnrichmentService do
  def enrich_batch(ideas) do
    idea_ids = Enum.map(ideas, fn idea -> idea[:idea_id] end)
    tags = fetch_tags(idea_ids)
    files = fetch_files(idea_ids)

    Enum.map(ideas, fn idea ->
      idea
      |> Map.put(:tags, Map.get(tags, idea[:idea_id], []))
      |> Map.put(:files, Map.get(files, idea[:idea_id], []))
    end)
  end

  defp fetch_tags(idea_ids) when idea_ids == [], do: %{}
  defp fetch_tags(idea_ids) do
    try do
      placeholders = Enum.map(idea_ids, fn _ -> "?" end) |> Enum.join(",")
      case IdeaBoard.Repo.query_maps(
             "SELECT it.idea_id, t.name AS tag_name FROM idea_tags it JOIN tags t ON t.tag_id = it.tag_id WHERE it.idea_id IN (#{placeholders})",
             idea_ids
           ) do
        {:ok, rows} ->
          (rows || [])
          |> Enum.group_by(fn map -> map[:idea_id] end, fn map -> %{name: map[:tag_name]} end)
        _ -> %{}
      end
    rescue
      _e in [DBConnection.ConnectionError, ArgumentError, MyXQL.Error] -> %{}
    end
  end

  defp fetch_files(idea_ids) when idea_ids == [], do: %{}
  defp fetch_files(idea_ids) do
    try do
      placeholders = Enum.map(idea_ids, fn _ -> "?" end) |> Enum.join(",")
      case IdeaBoard.Repo.query_maps(
             "SELECT idea_id, filename, original_name, mimetype FROM idea_files WHERE idea_id IN (#{placeholders})",
             idea_ids
           ) do
        {:ok, rows} ->
          (rows || [])
          |> Enum.group_by(fn map -> map[:idea_id] end, fn map ->
            %{filename: map[:filename], original_name: map[:original_name], mimetype: map[:mimetype]}
          end)
        _ -> %{}
      end
    rescue
      _e in [DBConnection.ConnectionError, ArgumentError, MyXQL.Error] -> %{}
    end
  end
end

File: lib/services/ideasEnrichmentService.js
/**
 * @fileoverview Anreicherungs-Utilities für Ideen: Batch-Laden von Tags, Dateien, Reaktionen und Live-Counters.
 * Vermeidet N+1-Abfragen durch Sammelabfragen.
 * @module lib/services/ideasEnrichmentService
 */
const db = require('../../config/db.js');

/**
 * Lädt Tags, Dateien und Nutzer-Reaktionen für eine Menge von Ideen (Batch).
 * @async
 * @param {Array<Object>} paginatedIdeas - Array mit Ideen-Objekten (muss `idea_id` enthalten).
 * @param {number} userId - Aktuelle Nutzer-ID (zur Erkennung eigener Likes/Dislikes).
 * @returns {Promise<Array<Object>>} Angereicherte Ideen-Objekte.
 */
async function enrichIdeasBatch(paginatedIdeas, userId) {
  if (!paginatedIdeas || paginatedIdeas.length === 0) return [];

  // IDs der sichtbaren Ideen extrahieren (für Batch-Abfragen)
  const visibleIds = paginatedIdeas.map(i => i.idea_id);
  const likesSet = new Set();
  const dislikesSet = new Set();
  const tagsMap = new Map();
  const filesMap = new Map();

  // Lade Live-Zähler aus der primären `ideas`-Tabelle, damit Statistiken sofort aktuell sind
  const liveCountsRows = await db.query(
    'SELECT idea_id, like_count, dislike_count, comment_count FROM ideas WHERE idea_id IN (?)',
    [visibleIds]
  );
  const liveCountsMap = new Map();
  for (const r of liveCountsRows) liveCountsMap.set(r.idea_id, r);

  const [tagsRows, filesRows, likedRows, dislikedRows] = await Promise.all([
    db.query(`-- sql
      SELECT l.idea_id, t.name
      FROM idea_tag_links l
      JOIN idea_tags t ON t.tag_id = l.tag_id
      WHERE l.idea_id IN (?)
    `, [visibleIds]),
    db.query(`-- sql
      SELECT file_id, idea_id, file_path, original_name
      FROM idea_files
      WHERE idea_id IN (?)
      ORDER BY uploaded_at DESC
    `, [visibleIds]),
    db.query('SELECT idea_id FROM likes WHERE user_id = ? AND idea_id IN (?)', [userId, visibleIds]),
    db.query('SELECT idea_id FROM dislikes WHERE user_id = ? AND idea_id IN (?)', [userId, visibleIds])
  ]);

  // Mappe die geladenen Tags pro Idee (idea_id -> [tagNames])
  tagsRows.forEach(row => {
    if (!tagsMap.has(row.idea_id)) tagsMap.set(row.idea_id, []);
    tagsMap.get(row.idea_id).push(row.name);
  });
  // Mappe die geladenen Dateien pro Idee (idea_id -> [fileObjs])
  filesRows.forEach(f => {
    if (!filesMap.has(f.idea_id)) filesMap.set(f.idea_id, []);
    filesMap.get(f.idea_id).push(f);
  });
  // Markiere Ideen, die vom aktuellen Nutzer geliked/disliked wurden
  likedRows.forEach(r => likesSet.add(r.idea_id));
  dislikedRows.forEach(r => dislikesSet.add(r.idea_id));

  // Baue die finalen angereicherten Idea-Objekte
  return paginatedIdeas.map(idea => {
    const live = liveCountsMap.get(idea.idea_id) || {};
    idea.like_count = Number(live.like_count ?? idea.like_count ?? 0) || 0;
    idea.dislike_count = Number(live.dislike_count ?? idea.dislike_count ?? 0) || 0;
    idea.comment_count = Number(live.comment_count ?? idea.comment_count ?? 0) || 0;
    idea.likes_count = idea.like_count;
    idea.dislikes_count = idea.dislike_count;
    idea.comments_count = idea.comment_count;
    idea.tags = tagsMap.get(idea.idea_id) || [];
    idea.files = filesMap.get(idea.idea_id) || [];
    idea.userLiked = likesSet.has(idea.idea_id);
    idea.userDisliked = dislikesSet.has(idea.idea_id);
    idea.isOwner = String(idea.user_id) === String(userId);
    idea.comments = [];
    return idea;
  });
}

/**
 * Normalisiert das Autoren-Feld: Wenn `author_username` vorliegt, kopiere es auf `author`.
 * @param {Array<Object>|Object|null} rows
 * @returns {Array<Object>|Object|null}
 */
function normalizeAuthor(rows) {
  if (!rows) return rows;
  if (Array.isArray(rows)) {
    for (const r of rows) {
      if (r && r.author_username !== undefined) {
        r.author = r.author_username;
        delete r.author_username;
      }
    }
    return rows;
  }
  if (rows.author_username !== undefined) {
    rows.author = rows.author_username;
    delete rows.author_username;
  }
  return rows;
}

module.exports = {
  enrichIdeasBatch,
  normalizeAuthor
};


